proposals/0015-constant-buffers.md

```diff
@@ -0,0 +1,380 @@
+@array = external addrspace(2) global [10 x i32]
+```
+
+Clang codegen will translate accesses to these global constants to `load`
+instruction from a pointer in `addrspace(2)`. These `load` instructions will be
+later replaced in an LLVM pass with constant buffer load intrinsic calls on a
+buffer resource handle. In order for the pass to generate the correct CBV loads
+it is going to need additinal information, such as which constants belong to
+which constant buffer, and layout of the buffer and any embedded structs.
+Codegen will generate this information as metadata.
+
+### Format of constant buffer metadata
+
+#### Mapping of constant global variables to constant buffer
+
+Clang codegen needs to emit metadata that will link `hlsl_constant` globals  to
+individual constant buffers. Metadata node for a single buffer will be a list of
+global variables where the first item will be the global variable represending
+the constant buffer. It will be followed by 1 or more `hlsl_constant` global
+variables that belong to this constant buffer, in the same order as they were
+declared in the source code.
+
+A named metadata node `hlsl.cbs` will then store a list all metadata nodes for
+individual buffers.
+
+For the HLSL code above the metadata will look like:
+
+```
+!hlsl.cbs = !{!1, !2}
+!1 = !{ptr @Constants.cb, ptr addrspace(2) @i, ptr addrspace(2) @s}
+!2 = !{ptr @OtherConstants.cb, ptr addrspace(2) @v, ptr addrspace(2) @array}
+```
+
+#### Layout information
+
+Clang also needs to include layout information for each constant buffer, and
+for any user defined structs  used in a constant buffer. Metadata node storing
+this information will include the name of the struct, its size, and then a list
+of offset for each field.
+
+A named metadata node `hlsl.layouts` will then capture a list of all layout
+metadata.
+
+For the HLSL code above the layout metadata will look like this:
+
+```
+!hlsl.layouts = !{!3, !4, !5}
+!3 = !{!"struct.S", i32 4, i32 0}                              ; size 4, element offsets 0
+!4 = !{!"struct.__layout_Constants", i32 20, i32 0, i32 16}    ; size 20, element offsets 0, 16
+!5 = !{!"struct.__layout_OtherConstants", i32 164, i32 0, i32 16} 
+                                        ; size 164, element offsets 0, 16
+```
+
+Since the module contains the struct definitions, we know which struct field is
+an array or not and how many subelements is contains. The only additional layout
+information that is not included here is the stride of the array elements. This
+can can be calculated as the size of the array element aligned to 16 (size of
+constant buffer row) because each array element always starts at the begining of
+a row.
+
+### Lowering to buffer load intrinsics
+
+A new pass `HLSLConstantAccess` will translate all `load` instructions in
+`hlsl_constant` address space to `llvm.{dx|spv}.resource.load.cbuffer`
+intrinsics. It will make use of the metadata generated by Clang codegen that
```

Comment:
What is the signature of this intrinsic? You can see how we generate code at https://godbolt.org/z/eq9E59WcM.

We need to the index of the member to generate the OpAccessChain. If the intrinsic contains the byte offset, then we will have to be able to translate the byte offset to member.

The ideal code pattern for SPIR-V is to access the CBuffers the same way we would access a StructuredBuffer. The only different would be the type for the handle. The index on the `resource_getpointer` would be the index of the member being accessed.

As I've mentioned in another comment, we would want the method of getting the offset for members of structuredBuffers and cbuffers to be the same.